### PR TITLE
ENH: add integrality constraints for linprog

### LINKED ISSUE #11673: intlinprog: integer linear program solver
**Is your feature request related to a problem? Please describe.**

``intlinprog``: a branch and bound mixed integer linear program solver built on top of ``scipy.optimize.linprog``.

@mdhaber has proposed possible integer constraints in #9269.  ILPs/MIPs are an entire class of problems that are very related to LPs, but which have an ecosystem of their own, suggesting that scipy might benefit from having a separate solver for ILPs/MIPs.  MATLAB and Octave also contain dedicated MIP solvers.

**Describe the solution you'd like**

A simple solver that can handle smaller to medium size problems.  Doesn't need to be CPLEX, Gurobi quality, something that will "just work" out of the box for reasonable problems and which has a scipy-consistent interface, i.e. works approximately like ``linprog.``

**Describe alternatives you've considered**

- https://pypi.org/project/mip/
- https://developers.google.com/optimization/mip/integer_opt

A lot of existing solutions use the optimization idioms of commercial solvers, i.e. create classes for Variables and Models, etc.  Other packages provide a wrapper on top of other solvers, no simple "out-of-the-box" experience.

**Additional context (e.g. screenshots)**

I have created a fork and added some code I was using for a personal project.  I've started to modify it to conform to the existing norms of the ``optimize`` module, but it still needs additional documentation and potentially a few upgrades as noted below:

- uses branch and bound
    - does not branch and cut, might perform better with a few cutting plane algorithms (e.g. Gomory cut)
- uses ``linprog`` `'revised simplex'` to solve relaxed linear program subproblems
    - `'interior-point'` leads to failure in some instances due to inherent approximations of the interior-point algorithm
- Implements several branching rules, easy to add new ones
    - does not implement a pseudo-cost branching rule currently
- Tries to implement a function signature similar to ``linprog``
- Works very well for small problems, medium problems (~8 variables) can be slow and depend a lot on search strategy used, i.e. depth-first, breadth-first, or best-first

The original work was done here: https://github.com/mckib2/bnb/blob/intprog/bnb/intlinprog.py

Scipy fork including ``intlinprog`` prototype: https://github.com/mckib2/scipy/blob/master/scipy/optimize/intlinprog.py

**Goal of this issue**

Get feedback and help on possible integration into scipy.

### PR DESCRIPTION
**Pull request by mckib2**
#### Reference issue

- gh-9269
- closes gh-11673
- gh-14455

#### What does this implement/fix?

- adds new `integrality` option to `scipy.optimize.linprog(method="highs")`
- interface tests
- tests with MIP problems
- extends the HiGHS Cython wrapper to expose MIP solver functionality

#### Additional information

`integrality` is an array of variable type constraints corresponding to decision variables. If left as `None` (default), all variable types are assumed to be continuous (i.e., simple linear program).  Three more types are available:
- Integer variable
- Semi-continuous variable (decision variable must be within `bounds` or take value ``0``)
- Semi-integer variable (decision variable must be an integer within `bounds` or take value ``0``)

### CODE REVIEW

File: scipy/optimize/tests/test_linprog.py [hunk 1]
```diff
@@ -754,7 +772,7 @@ def test_zero_column_2(self):
                       method=self.method, options=self.options)
         _assert_unbounded(res)
         # Unboundedness detected in presolve
-        if self.options.get('presolve', True):
+        if self.options.get('presolve', True) and "highs" not in self.method:
```
Comment (mckib2) [inline scipy/optimize/tests/test_linprog.py:775]: It appears that HiGHS presolve may be correct here and SciPy is wrong.  @mdhaber can you take a look at https://github.com/ERGO-Code/HiGHS/issues/648 and say if we think scipy presolve is deducing unboundedness too soon?

Comment (mdhaber) [reply scipy/optimize/tests/test_linprog.py:775]: There message the user will get in this case is:
> If feasible, the problem is (trivially) unbounded due to a zero column in the constraint matrices. If you wish to check whether the problem is infeasible, turn presolve off.

So I guess we are doing what a lot of solvers do, reporting "infeasible or unbounded", in a sense.

Comment (mckib2) [reply scipy/optimize/tests/test_linprog.py:775]: Okay, so it looks correct to exclude highs' methods this check.  Does this need a note or is it good as is?

Comment (mdhaber) [inline scipy/optimize/tests/test_linprog.py:775]:
```suggestion
        if self.options.get('presolve', True) and "highs" not in self.method:
        # HiGHS detects unboundedness or infeasibility in presolve
        # It needs an iteration of simplex to be sure of unboundedness
        # Other solvers report that the problem is unbounded if feasible
```

File: scipy/optimize/_linprog_highs.py [hunk 2]
```diff
@@ -236,6 +240,20 @@ def _linprog_highs(lp, solver, time_limit=None, presolve=True,
                     function with respect to the lower and upper
                     `bounds`.
 
+            mip_node_count : int
+                Only present when `integrality` is not `None`. The number of
+                subproblems or "nodes" solved by the MILP solver.
```
Comment (mdhaber) [inline scipy/optimize/_linprog_highs.py:245]:
```suggestion
                The number of subproblems or "nodes" solved by the MILP
                solver. Only present when `integrality` is not `None`.
```
I think we should lead with the description and follow with the details.

File: scipy/optimize/_linprog_highs.py [hunk 3]
```diff
@@ -236,6 +240,20 @@ def _linprog_highs(lp, solver, time_limit=None, presolve=True,
...
+
+            mip_dual_bound : float
+                Only present when `integrality` is not `None`. The MILP
+                solver's final estimate of the lower bound on the optimal
+                solution.
```
Comment (mdhaber) [inline scipy/optimize/_linprog_highs.py:250]:
```suggestion
                The MILP solver's final estimate of the lower bound on the
                optimal solution. Only present when `integrality` is not
                `None`.
```

File: scipy/optimize/_linprog_highs.py [hunk 4]
```diff
@@ -236,6 +240,20 @@ def _linprog_highs(lp, solver, time_limit=None, presolve=True,
...
+
+            mip_gap : float
+                Only present when `integrality` is not `None`. The difference
+                between the final objective function value and the final dual
+                bound.
```
Comment (mdhaber) [inline scipy/optimize/_linprog_highs.py:255]:
```suggestion
                The difference between the final objective function value
                and the final dual bound. Only present when `integrality`
                is not `None`.
```

File: scipy/optimize/_linprog_highs.py (hunk 2 above)
Comment (mckib2) [reply scipy/optimize/_linprog_highs.py]: Done

**Comment by mdhaber** [comment]
I'll go ahead and merge this on 1/7 unless I hear back from anyone before then.
Then, I'll open up mckib2/scipy#31 on the SciPy repo and email the mailing list about that.

**Pull request merged**